internal/merger/merger.go

```diff
@@ -0,0 +1,37 @@
+// Package merger contains code for merging channels.
+package merger
+
+import (
+	"sync"
+
+	"github.com/m-lab/ndt7-client-go/spec"
+)
+
+// Merge merges the content of two channels into the returned channel.
+func Merge(leftch, rightch <-chan spec.Measurement) <-chan spec.Measurement {
```

Comment:
Please pass in a `ctx context.Context`, and make each `for` loop exit when the context is canceled.  I suggest doing:

```go
 ok := true
 for ok {
  select {
  case m := <- leftch:
    if m == nil {
      ok = false
    } else {
      out <- m
    }
  case <-ctx.Done():
    ok = false
  }
}
```

Then context cancellation will also cause the `Merge` goroutines to exit. Guaranteeing context cancellation is easier than guaranteeing that channels will be closed. Channel closure is actually remarkably subtle - only writers should close channels, never readers, and only one writer. If a channel is never explicitly closed, but both its endpoints pass out of scope, then the channel will be garbage collected and closed.

Reply:
UGH. I wrote a much better comment here and it got eaten. I suggest the following code for Merge:

```go
func Merge(ctx context.Context, chs... <-chan spec.Measurement) <-chan spec.Measurement {
        outch := make(chan spec.Measurement)
￼       var wg sync.WaitGroup
￼       wg.Add(len(chs))
        for _, ch := range chs {
                go func(ch <-chan spec.Measurement) {
                         ok := true
                         for ok {
                                 select {
                                 case m := <-ch:
                                         if m == nil {
                                                 ok = false
                                         } else {
                                                 outch <- m
                                         }
                                 case <-ctx.Done():
                                         ok = false
                                 }
                         }
                         wg.Done()

                }(ch)
        }
        go func() {
￼               wg.Wait()
￼               close(outch)
￼       }()
￼       return outch
}

```